### PR TITLE
refactor(api): Create robot_server sub project

### LINKED ISSUE #5010: Create an opentrons python library sub-project in mono repo
## overview

It can me named py-opentrons or something else more imaginative.
This would contain everything in api project minus opentrons.server and opentrons.app packages. Almost everything in current api is applicable to building a package vs an application.

~~- [ ] new sub project created (was decided to keep API name)~~
~~- [ ] all sources moved from `api` to new sub project minus webservice packages.~~
- [x] move webservice code out of package
- [x] pipfile loses aiohttp and fastapi dependencies
- [x] webservice documentation is removed

### LINKED ISSUE #5011: Create api-server sub-project
## overview
This is the api server application. For the time being both the aiohttp (opentrons.server) and fastapi (opentrons.app) packages will be in this project.
Directory Structure
```
api
| README.md
| Makefile
| ...
└- tests
|  | __init__.py
|  └- app
└- app
   | __init__.py
   | main.py
   └- aiohttp ( to be deprecated )
   └- fastapi ( move contents one level up when aiohttp is deprecated )
```

- [x] directory structure is created
- [x] create pipfile 
- [ ] move webservice related documentation from `api` to new `api-server`

### PR DESCRIPTION
**Pull request by amitlissack**
## overview

This supersedes #5055 

api is being split up into two projects. This creates the webservice portion of the split. As well as some small changes to the existing `api` project.

This differs from #5055 in that we do not move the `opentrons.server` package to `robot_server`. Instead, if configured to start the aiohttp api server we simply import  `opentrons.server.run` to start the application.

The benefit is that we preserve git history. Thanks @sfoster1 for idea.

Other small changes are some attempts to kill threads in deck calibration and the MainRouter.

## review requests

This is a WIP. Please do not bother with the `robot_server` Makefile and buildroot files. Changes to those will come in a future PR.

closes #5011, closes #5010

### CODE REVIEW

**Comment by mcous** [comment]
Looks like you're want to add a new entry to the JS lint ignore to avoid existing formatting of the `release-notes.md` from failing the at the lint stage

```ignore
# .eslintignore
api/**
update-server/**
robot-server/**
# ...
```

File: api/src/opentrons/create_hardware.py [hunk 1]
```diff
@@ -122,85 +118,33 @@ def initialize_robot(loop, hardware):
     log.info(f"Name: {name()}")
 
 
-def run(hardware, **kwargs):  # noqa(C901)
+async def create_hardware(hardware_server: bool = False,
```
Comment (amitlissack) [inline api/src/opentrons/create_hardware.py]: `robot-server` calls this function at startup to create the hardware API instance. The hardware instance is supplied to the aiohttp or fastapi run method.

File: api/src/opentrons/create_hardware.py [hunk 2]
```diff
@@ -122,85 +118,33 @@ def initialize_robot(loop, hardware):
...
-                "Hardware server requested but apiv1 selected, not starting")
-    server.run(
-        hardware,
-        kwargs.get('hostname'),
-        kwargs.get('port'),
-        kwargs.get('path'),
-        loop)
-
-
-def main():
```
Comment (amitlissack) [inline api/src/opentrons/create_hardware.py]: There is no http server entrypoint in `api` project anymore. The `opentrons.server` package is imported by robot_server.

File: api/src/opentrons/server/__init__.py [hunk 3]
```diff
@@ -100,12 +100,15 @@ def init(hardware: 'HardwareAPILike' = None,
     app = web.Application(middlewares=[error_middleware])
     app['com.opentrons.hardware'] = hardware
     app['com.opentrons.motion_lock'] = ThreadedAsyncLock()
-    app['com.opentrons.rpc'] = RPCServer(
-        app, MainRouter(
-            hardware, lock=app['com.opentrons.motion_lock'], loop=loop))
     app['com.opentrons.response_file_tempdir'] = tempfile.mkdtemp()
     app['com.opentrons.http'] = HTTPServer(app, CONFIG['log_dir'])
 
+    main_router = MainRouter(hardware, lock=app['com.opentrons.motion_lock'],
```
Comment (amitlissack) [inline api/src/opentrons/server/__init__.py]: This is a change from edge. We hold on to the `MainRouter` in order to clean up in an aiohttp shutdown handler. py37 does not terminate threads the same way as py36.

File: robot-server/robot_server/main.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+             "because extra dependencies are required.")
+    arg_parser.add_argument(
+        '--hardware-server-socket', action='store',
+        default='/var/run/opentrons-hardware.sock',
+        help='Override for the hardware server socket')
+    return arg_parser
+
+
+def run(hardware: HardwareAPILike,
+        hostname,
```
Comment (sfoster1) [inline robot-server/robot_server/main.py]: these should be type `Optional[str]` right

File: robot-server/robot_server/main.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        port,
+        path: typing.Optional[str] = None):
```
Comment (sfoster1) [inline robot-server/robot_server/main.py]: `path: str = None` should get automatically changed to `typing.Optional[str]`, you only need the `Optional` if you're not defaulting to `None`

File: api/src/opentrons/deck_calibration/endpoints.py [hunk 6]
```diff
@@ -13,14 +13,35 @@
...
-from . import jog, position, dots_set, z_pos
+from opentrons.deck_calibration import jog, position, dots_set, z_pos
 from opentrons.util.linal import add_z, solve, identity_deck_transform
 
-session = None
 mount_by_name = {'left': Mount.LEFT, 'right': Mount.RIGHT}
 log = logging.getLogger(__name__)
 
 
+class SessionWrapper:
```
Comment (amitlissack) [inline api/src/opentrons/deck_calibration/endpoints.py]: SessionWrapper replaces global session. Allows doing some cleanup when creating a new session

File: robot-server/release-notes.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
+# Robot OS Changes from 3.14.1 to 3.15.2
```
Comment (mcous) [inline robot-server/release-notes.md]: Might make sense to pull the latest version of this file in from `edge` (or leave it out of this PR for now)

File: robot-server/.coveragerc [hunk 8]
```diff
@@ -0,0 +1,3 @@
+[paths]
+source =
+    src/opentrons
```
Comment (mcous) [inline robot-server/.coveragerc]: This (and a few others) looks like it might be copypasta. Cool with me to punt whatever to the Makefile PR, just pointing it out from an overabundance of caution

File: robot-server/pypi-readme.rst [hunk 9]
```diff
@@ -0,0 +1,51 @@
+.. _Full API Documentation: http://docs.opentrons.com
```
Comment (mcous) [inline robot-server/pypi-readme.rst]: Is this file needed?

File: robot-server/robot_server/main.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    arg_parser.add_argument(
+        "-H", "--hostname",
+        help="TCP/IP hostname to serve on (default: %(default)r)",
+        default="localhost"
+    )
+    arg_parser.add_argument(
+        "-P", "--port",
+        help="TCP/IP port to serve on (default: %(default)r)",
+        type=int,
+        default="8080"
```
Comment (mcous) [inline robot-server/robot_server/main.py]: Most of our front-end machinery is expecting any robot server (including dev servers) to be on port 31950. Should we change this value?

File: robot-server/robot_server/package.json [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "@opentrons/robot-server",
+  "version": "3.15.2",
```
Comment (mcous) [inline robot-server/robot_server/package.json]: Latest version in edge is 3.16.1; might want to change this to avoid any weird conflicts.

We need to have a larger discussion about versioning, and we've also got time on the roadmap to overhaul the versioning / release process, so this comment is geared towards "making sure nothing weird happens with our existing machinery" rather than "the existing machinery is good" 🙃

File: robot-server/release-notes.md (hunk 7 above)
Comment (amitlissack) [reply robot-server/release-notes.md]: I'm gonna remove this file. It's not applicable to robot server. Whatcha think?

File: robot-server/pypi-readme.rst (hunk 9 above)
Comment (amitlissack) [reply robot-server/pypi-readme.rst]: It is not.
